src/util/parallel/data.ml

```diff
@@ -0,0 +1,242 @@
+        let bucket = Array.get buckets (hash mod (Atomic.get hm.size)) in
+        match Atomic.get bucket with
+        | None ->
+          let new_bucket = Bucket.create key in
+          let success = Atomic.compare_and_set bucket None (Some new_bucket) in
+          if success then (
+            (new_bucket.value, true) 
+          )
+          else find_create_inner hm key hash buckets 
+        | Some bucket -> 
+          let value, was_created = Bucket.find_create bucket key in
+          (value, was_created)
+      in
+      let current_generation = Atomic.get hm.resize_generation in
+      let hash = abs @@ H.hash key in
+      let value, was_created = find_create_inner hm key hash (Atomic.get hm.buckets) in
+      if (current_generation mod 2 == 0) && (Atomic.get hm.resize_generation == current_generation || not was_created) then (
+        if (Atomic.get hm.nr_elements >= Atomic.get hm.size * 2) then (
+          resize hm;
+        );
+        if (was_created) then Atomic.incr hm.nr_elements;
+        (value, was_created)
+      )
+      else (
+        while (Atomic.get hm.resize_generation == current_generation) do () (* spin *)
+        done;
+        find_create hm key)
+
+
+    and resize hm =
+      let current_generation = Atomic.get hm.resize_generation in
+      if ((current_generation mod 2 == 0) && Atomic.compare_and_set hm.resize_generation current_generation (current_generation+1)) then (
+
+        let old_size = Atomic.get hm.size in
+        let new_size = old_size * 2 in
+
+        let new_buckets = Array.init new_size (fun _ -> Atomic.make None) in
+        let rec rehash_bucket (bucket: Bucket.t option Atomic.t) =
+          match Atomic.get bucket with
+          | None -> ()
+          | Some element -> 
+            begin
+              let value = element.value in
+              let key = element.key in
+              let hash = abs @@ H.hash key in
+              let new_location = Array.get new_buckets (hash mod new_size) in
+              let _ = match Atomic.get new_location with
+                | None -> (ignore @@ Atomic.set new_location (Some (Bucket.create_with_value key value)))
+                | Some new_bucket -> 
+                  let newer_bucket = Bucket.create_with_value_and_next key value new_bucket in
+                  (ignore @@ Atomic.set new_location (Some newer_bucket)) in
+              rehash_bucket element.next
+            end
+        in
+        Array.iter (fun bucket -> rehash_bucket bucket) (Atomic.get hm.buckets);
+
+        Atomic.set hm.buckets new_buckets;
+        Atomic.set hm.size new_size;
+        Atomic.incr hm.resize_generation;
+      )
+
+    let to_value_seq hm =
+      let bucket_seq = Array.to_seq (Atomic.get hm.buckets) in 
+      let rec bucket_to_value_seq (bucket : Bucket.t option Atomic.t) = match Atomic.get bucket with
+        | None -> fun () -> Seq.Nil
+        | Some b -> fun () -> Seq.Cons (b.value, bucket_to_value_seq b.next) in
```

Comment:
Why is `fun () ->` needed here?

Reply:
This is how you build a Seq. Replaced `fun () -> Seq.Nil` with `Seq.empty` for more clarity.

(In the bucket above, this is achieved with the `()` parameter in `aux`. The approach is useful here, because the bucket itself can be `None`)